### DISCUSSION TITLE
The `file` command feedback

### DISCUSSION THREAD

**Discussion by andrii-bodnar** [crowdin/crowdin-cli#713] [Show and tell]
The [3.17.0](https://github.com/crowdin/crowdin-cli/releases/tag/3.17.0) release brings the much-requested command to manage files in the Crowdin project individually and independently from the configuration file.

The [`file` command](https://crowdin.github.io/crowdin-cli/commands/crowdin-file) is a low-level command that allows you to manage your files with more flexibility and control. This command also supports Crowdin's new String-based projects :rocket:

Visit the [Files Management](https://crowdin.github.io/crowdin-cli/tutorials/files-management) tutorial to learn more about the new commands and how they compare to the existing approaches.

⚠️ This command is being released with an "experimental" flag, which means that something may change in the future (based on the feedback we get).

**So your feedback is highly appreciated!** ✨

**Comment by wes-nz** [comment]
I initially uploaded my translations manually via the website. I then decided to automate it via the `file upload` command. 
I noticed that my strings with newline characters `\n` were processed differently on the editor. 

If it is upload via the website the strings are kept raw
<img width="182" alt="Screenshot 2024-02-03 at 5 14 52 pm" src="https://github.com/crowdin/crowdin-cli/assets/2945380/19d45214-6b83-4aa8-8cae-113f16200fc9">
but if it is via the `file upload` command it is not
<img width="165" alt="Screenshot 2024-02-03 at 5 15 18 pm" src="https://github.com/crowdin/crowdin-cli/assets/2945380/fcb319f6-08ba-4381-a98f-0997b35cb33a">

I can't see any option on `file upload` related to that problem, let me know if I am missing something.
Thanks!

**Reply by andrii-bodnar** [reply]
@wes-nz Thanks for the feedback! Could you please provide a sample file so we can easily reproduce this on our end? Thank you!

**Reply by wes-nz** [reply]
`strings.xml` (Android) XML uploaded via the website
```XML
<?xml version="1.0" encoding="UTF-8"?>
<resources>
    <string name="welcome_title_1">\nHello World</string>
</resources>
```

`Localizable.strings` (iOS) uploaded via the `file upload`
```Strings
"welcome_title_1" = "\nHello World";
```

**Reply by andrii-bodnar** [reply]
@wes-nz I just tested these examples and the upload result is the same whether I use the UI or the CLI.

**Reply by wes-nz** [reply]
@andrii-bodnar weird, are you on macOS? I just created a fresh project, upload the example `strings.xml`, then linked the project with `crowdin init`, followed by `crowdin file upload ./Localizable.strings -b main` I end up with two translations for the same key.
<img width="446" alt="Screenshot 2024-02-09 at 8 24 16 am" src="https://github.com/crowdin/crowdin-cli/assets/2945380/3db5f327-57b0-4a29-a5d7-43978ffe5e71">
Let me know if you have any questions.

**Reply by andrii-bodnar** [reply]
@wes-nz yes, I'm on macOS. Could you please attach the file?

**Reply by wes-nz** [reply]
[XML+Strings.zip](https://github.com/crowdin/crowdin-cli/files/14214595/XML%2BStrings.zip)

**Reply by andrii-bodnar** [reply]
@wes-nz thanks a lot! I was able to reproduce the problem, it's an API side issue. Already reported to the team.

**Comment by jimmysafe** [comment]
Hello! I am trying to use this very nice new feature of yours and i ran into an issue:

We are using Crowdin for Figma plugin and we also have a webapp where devs push new source strings as well.
When we add a new string through Crowdin for Figma plugin it gets uploaded ok:

<img width="1002" alt="Screenshot 2024-03-28 at 17 15 12" src="https://github.com/crowdin/crowdin-cli/assets/12184812/e43a5ddf-de82-42a1-856a-541e8d19ccef">

However when devs try and use the crowdin cli to update that file with new source strings, the whole file gets reset and it gets only the newly pushed sources by devs.

here is the command we run:

`npx crowdin file upload ./locales/it/signup-mobile.json -d /mobile/signup-mobile.json -b master --no-cleanup-mode`

What we expect is that we see all the changes devs made to the file **PLUS** the string we pushed from figma, but unfortunately it gets deleted in the process.

Is there any way to upload a file with this new command and only update it with new source strings ?
If not, is this something you guys are planning ?

Thanks in advance, keep up the great work 🫡

**Reply by andrii-bodnar** [reply]
Hi @jimmysafe, thanks for the feedback, it's very valuable for us!

This is expected behavior in file-based projects. The thing here is that pushing strings from Figma uses different API (Add String) and it modifies the file in Crowdin by adding these strings. The developers may not have these changes locally and when they push the file to the Crowdin project it overwrites the file state there as it updates the entire file.

I would recommend using [string-based](https://support.crowdin.com/creating-project) projects for your workflow. By the way, the `--no-cleanup-mode` only works for this type of project.

> In string-based projects, the focus is on managing translatable content as individual strings rather than source files. In this project type, after uploading source files, Crowdin parses them into source strings. Unlike file-based projects, source files aren’t stored, and the emphasis is on managing the content at the string level. This approach is beneficial when dealing with projects that involve continuous content updates, dynamic content, or where a string-oriented structure is preferred.

In this case, you'll need to use [Bundles](https://support.crowdin.com/bundles-sb/) to download the translations from Crowdin. As you use the JSON file format, you'll need to use the [JSON String Exporter](https://store.crowdin.com/json-string-exporter-v2) in the Bundle configuration.
